Add Table of Contents block (dynamic rendering + hooks version)

ZebulanStanphill opened the pull request:
## Description
Closes #11047. Related: #22874.

This is a derivative of #21040 (which itself is a derivative of #15426) that has been massively refactored to support h1-h6 elements anywhere in the content (not just core Heading blocks), as well as paginated posts.

Big props to @SeanDS, @ashwin-pc, [sorta brilliant](https://sortabrilliant.com/guidepost/), and everyone else who has helped make this block a reality! :smile:

## Screenshots

![image](https://user-images.githubusercontent.com/19592990/106762504-83f4fe80-65fb-11eb-8c4f-80fae91c9ecb.png)

![image](https://user-images.githubusercontent.com/19592990/106762955-fbc32900-65fb-11eb-96d7-0a6298bc1c72.png)

![image](https://user-images.githubusercontent.com/19592990/106762365-632ca900-65fb-11eb-9085-2dbadab32f17.png)

## Implementation notes

There are some notable quirks about the implementation I should point out:

First, both the front-end and editor implementations creates temporary HTML documents (or unattached `<div>`s in the case of the editor) in memory during the render process, just so they can be filled with the current post's content and parsed using DOM APIs to determine how many `h1`-`h6` tags and `<!--nextpage-->` comments there are in the current post. This feels hacky and sounds like it could be a performance issue, but in practice I haven't noticed any serious performance impact, and I am not aware of a better solution.

Second, when using the "Only include current page" option with a paginated post, the editor implementation is technically not able to determine which page a Table of Contents block will appear on with 100% accuracy. It only accounts for `<!--nextpage-->` comments that occur in a Page Break or Classic block. The front-end implementation does not have this imperfection.

To fix this using the currently available APIs would require searching the content of every block preceding the Table of Contents block **every time the editor content changes**. As it is, it currently only scans the content of Classic blocks, and the content of Page Break blocks don't have to be scanned since, well, just knowing they are a Page Break block already tells you all you need to know.

However, I doubt this slight inconsistency would be an issue in practice. I can't think of a situation where someone wouldn't just use the Page Break block when creating new content, and old content will all be in the Classic block. It would be trivial to add support for page breaks within Custom HTML blocks, but that seems like an anti-pattern to me.

It's also worth noting that I scan the Classic block's content for page breaks by reading its `content` attribute. It's not possible to support every block through this method since 3rd party blocks could use a differently-named attribute to store their text/html content, and I can't predict what the content attribute will be called. To support any given block would require getting the blocks' inner HTML... and I'm not aware of any API to do that from the editor.

Technical feedback on any of this is very welcome. There has been some discussion in #22874 about a document-outline/headings API to provide a method of getting the required data without constantly scanning raw HTML for different elements/comments, but I think this PR in its current state may be good enough as-is.

## Todo

- [x] This block needs an icon designed for it.
- [x] Determine whether this block should be called "Table of Contents", "Document Outline", "Outline", or something else.
- [x] There is currently an issue where clicking the block to select it doesn't work in some cases. It doesn't appear to be caused by this PR, but it should probably be solved prior to merge.

SeanDS committed:
SeanDS's implementation #21040 (https://github.com/WordPress/gutenberg/pull/21040).

Copy changes from pull request #15426 (https://github.com/WordPress/gutenberg/pull/15426).

Adds Table of Contents block to the editor.

Code contributions in this commit entirely made by ashwin-pc, originally based on the
"Guidepost" block by sorta brilliant (https://sortabrilliant.com/guidepost/).

Apply polish suggestions from code review.

Improve variable names.

Add comment

Get rid of autosync (users should now convert to list if they want to edit the contents)

Add ability to transform into list; remove unused ListLevel props

Update table-of-contents block test configuration

Simplify expression

Remove unused function

Remove unused styles.

Rename TOCEdit to TableOfContentsEdit

Apply suggestions from code review

Remove non-existent import

Make imports explicit

Remove unused function

Change unsubscribe function to class property

Change JSON.stringify comparison to Lodash's isEqual

Turns out refresh() is required

Remove unnecessary state setting

Don't change state on save

Change behaviour to only add links if there are anchors specified by the user

Newline

Replace anchor with explicit key in map since anchor can now sometimes be empty

Update test data

Update packages/block-library/src/table-of-contents/block.json

Rename ListLevel to ListItem for clarity and polish.

Co-authored-by: ashwin-pc <ashwinpc1993@gmail.com>
Co-authored-by: Daniel Richards <daniel.p.richards@gmail.com>
Co-authored-by: Zebulan Stanphill <zebulanstanphill@protonmail.com>
Co-authored-by: JR Tashjian <jrtashjian@gmail.com>

ZebulanStanphill committed:
Polish, use hooks, and half-fix undo behavior.

ZebulanStanphill commented:
I have no idea what's causing the test failure.

SeanDS commented:
Looks like undo doesn't work. I set up a hierarchy of headers and then deleted one, saw the contents update, then pressed undo and it didn't come back again (it seems to quickly appear then get deleted again). Maybe the `componentDidUpdate` function you removed was important for this.

ZebulanStanphill commented:
I've spent some time trying to figure out how to fix the undo issue, and now it's working better. You can now undo, but there are two undo steps: one for the Heading block and one for the Table of Contents block. If you press Undo only once, it changes the Table of Contents but not the Heading. This actually seems to be an improvement over #21040, where 3 undo steps are created.

Ideally, there would be only one undo step to prevent the Table of Contents from becoming out-of-sync, but I have no idea how to go about solving this.

@ellatrix, if I recall correctly, you have some experience working on the undo system. Any ideas?

ZebulanStanphill commented:
Notably, the same test that was failing earlier is still failing.

paaljoachim commented:
I'll also let Andrew @aduth and Riad @youknowriad know about your rubix cube puzzle, as they might have some advice..:)

aduth commented:
This block is rather unique in how its own output is dependent on values derived from other blocks. But I'm not sure if this needs to be (or makes sense as) the attributes of this block. This partly ties into the issues around Undo, since it's the redundancy of the changes in blocks state which is causing the extra Undo (the original block's change, then the sync).

Even though it can often be generated statically based on the other block's currently in the same post, the fact that it's not self-contained—but rather relies on external circumstances—leads me to think it falls more under the definition of a [dynamic block](https://developer.wordpress.org/block-editor/tutorials/block-tutorial/creating-dynamic-blocks/). There could optionally be some static fallback markup based on the last known markup of the post, but it leads to some challenges in reconciling potential differences in future edits (block invalidations).

I'd be receptive to challenges, but I might approach this block differently:

- No attributes of its own.
- The `edit.js` implementation can look similar to how it does currently, except it sets its own internal _component_ state, rather than `setAttributes`. It becomes more of a local cache of its rendered output, but not persisted attributes. It could even be computed on-the-fly if it were performant enough. I'd worry (for longer posts especially) that some amount of caching would be appropriate.
- `save.js` always returns `null`.
   - Again, not ideal from the perspective of what the block should be capable of outputting, but a pragmatic compromise based on what's currently possible to reconcile for block validation.
- As a [dynamic block](https://developer.wordpress.org/block-editor/tutorials/block-tutorial/creating-dynamic-blocks/), there's a server-side `render_callback` which generates the markup based on other blocks in the current post when shown on the front-end.

cc @youknowriad

ZebulanStanphill commented:
@aduth I've almost finished refactoring the block to use dynamic PHP rendering, but I've run into an issue. I'm not sure how to get the current post in the PHP code. I know there's a way to get the post id in JavaScript and from there I could save it to a block attribute and reference it in the PHP render callback; however, I would prefer if I could avoid saving the post id to the block.

Currently, the PHP render callback always renders an empty string, so the block only appears in the editor (where the `edit.js` implementation is used).

aduth commented:
I expect the `$post` global could likely be available within the context of a `render_callback`.

You may also be interested in some of the ongoing effort of #21467 (#19685), which seeks to address this in a more standardized manner.

ZebulanStanphill commented:
@aduth
>I expect the $post global could likely be available within the context of a render_callback.

That's what I thought, but I've tried it, and it doesn't seem to work. Let me know if you spot any errors in my code.

I'll keep an eye on #21467.

ZebulanStanphill committed:
Make dynamic, and add support for paginated posts + h1-h6 tags outside of core blocks.

DRY out usage of list item class name.

Improve handling of heading attributes.

Add proper placeholder state in editor.

Fix mistakes, improve naming/description, and polish code.

Add support for page breaks in a Classic block.

Always ignore empty headings.

Various performance improvements.

Change List block conversion to be a button in the toolbar. (It can't use the transform API since it uses dynamic data.)

Remove unused key from hierarchical heading list.

aduth commented on packages/block-library/src/table-of-contents/index.php:
```diff
@@ -0,0 +1,223 @@
+ * @return array The list of Heading blocks.
+ */
+function block_core_table_of_contents_get_heading_blocks() {
+	$blocks = parse_blocks( gutenberg_get_post_from_context()->post_content );
```
To clarify, regarding https://github.com/WordPress/gutenberg/pull/21234#issuecomment-611313837, you had tried this without success?

```suggestion
	global $post;
	$blocks = parse_blocks( $post->post_content );
```

ZebulanStanphill replied:
Yeah, I tried that and it didn't work. I also tried this and it didn't work either:
```suggestion
	$blocks = parse_blocks( get_post()->post_content );
```

aduth commented:
@ZebulanStanphill I tested this locally. One thing I immediately noticed is that the block wasn't being registered at all, so the PHP code wasn't actually being run. I fixed this in 35efbbf. At that point, I tried using the `$post` global directly. It seemed to work fine for me? I pushed this change in 0548b86. Another error I saw was with the `array_map` call in `gutenberg_block_core_table_of_contents_blocks_to_heading_list`, having the arguments reversed (fixed in 3664521).

There's still an issue though, since if you have error logging enabled in your development environment (should be configured by default in the [standard environment](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/getting-started.md)), there's still a warning:

```
Notice: Undefined index: level in /var/www/html/wp-content/plugins/gutenberg/build/block-library/blocks/table-of-contents.php on line 38
```

It's an interesting challenge. Heading does have a default value:

https://github.com/WordPress/gutenberg/blob/3130e85a217fc376102baaad18fdb6e07e1cb9c3/packages/block-library/src/heading/block.json#L14-L17

So one might think it's safe to assume this would always be assigned. But since you're [using `parse_blocks` directly](https://github.com/WordPress/gutenberg/blob/3664521304cecfea7f9639778490337750a6169f/packages/block-library/src/table-of-contents/index.php#L16), these default attributes aren't assigned. They become assigned as [part of `WP_Block_Type::prepare_attributes_for_render`](https://github.com/WordPress/wordpress-develop/blob/eca3d5a2f331c470d5d248fe45a6bae0d9224b84/src/wp-includes/class-wp-block-type.php#L124-L169), which is never called in this flow.

Some options to consider could be:

- Find a way to get `prepare_attributes_for_render` to be called. The [`render_block`](https://github.com/WordPress/wordpress-develop/blob/eca3d5a2f331c470d5d248fe45a6bae0d9224b84/src/wp-includes/blocks.php#L434-L500) implementation could be useful for reference.
- At a framework level, we could revisit whether it makes sense for defaults to be applied at a different step in the parsing process (i.e. as part of the parse). I'm not entirely convinced this is the right direction to consider.
- Recreate the effects of `prepare_attributes_for_render` in assigning the default value, or have some fallback behavior for when the level isn't assigned. I'd be careful to avoid _assuming_ anything about that default though (including specific values or even if a default is assigned at all).

ZebulanStanphill commented:
@aduth Thanks a lot for the fixes! I should've remembered to check and see if the block was actually being registered server-side in the first place.

I was using `wp-env` for my testing a little while back, but an update broke it for me on Linux, and I haven't been able to get it working since then, so I've been using [gutenberg.run](http://gutenberg.run/) to test my branches, which doesn't have `WP_DEBUG` enabled, if I recall correctly.

`WP_Block_Type::prepare_attributes_for_render` is called by [`WP_Block_Type::render`](https://github.com/WordPress/wordpress-develop/blob/2ae5bb324d1644eebdd8110b92923a7199d2e101/src/wp-includes/class-wp-block-type.php#L102). However, it is only called for non-dynamic blocks. I guess this means default values in `block.json` are just ignored by default for dynamic blocks? That seems like a bug to me.

Additionally, in the case of the Heading block, it doesn't make sense to have a heading with an "auto" heading level, as far as I am aware. I think the Heading block should be updated to always explicitly set the level attribute upon insertion. What do you think?

One more thing: how do we support headings that aren't in a core Heading block? I know several plugins add their own "Advanced Heading" kind of block, and some Accordion blocks have the option use HTML heading elements for their labels. Should we switch to searching the post content for all HTML heading elements? That would make the undefined Heading block level attribute issue irrelevant, but the different approach would probably introduce some new challenges to overcome as well. What do you think?

aduth commented:
> `WP_Block_Type::prepare_attributes_for_render` is called by [`WP_Block_Type::render`](https://github.com/WordPress/wordpress-develop/blob/2ae5bb324d1644eebdd8110b92923a7199d2e101/src/wp-includes/class-wp-block-type.php#L102). However, it is only called for non-dynamic blocks. I guess this means default values in `block.json` are just ignored by default for dynamic blocks? That seems like a bug to me.

Did you mean the reverse of this? It is only called for dynamic blocks, and default values are not assigned for static blocks. I suspect it was likely done this way intentionally, since a static block doesn't require any further server-side processing.

> Additionally, in the case of the Heading block, it doesn't make sense to have a heading with an "auto" heading level, as far as I am aware. I think the Heading block should be updated to always explicitly set the level attribute upon insertion. What do you think?

I feel as though I've seen this idea discussed at some point in the past. It does seem like a reasonable enhancement, sure. There is #20870 which explores _hints_ to correct possible errors (I seen you've commented there as well). Smart defaults could complement this.

I assume it would be a separate effort to this one, though.

> One more thing: how do we support headings that aren't in a core Heading block? I know several plugins add their own "Advanced Heading" kind of block, and some Accordion blocks have the option use HTML heading elements for their labels. Should we switch to searching the post content for all HTML heading elements? That would make the undefined Heading block level attribute issue irrelevant, but the different approach would probably introduce some new challenges to overcome as well. What do you think?

It's a good question. Have you looked to see how the "Document Outline" considers headings? Not to say it couldn't be a good idea to explore being more open to heading tags used in non-core blocks, but it would give some idea for prior art. I'd think the two ought to be consistent in any case. One worry I might have with detecting headings anywhere is if third-party blocks may be using heading tags in ways which don't necessarily fit in the model of a "Document Outline" or "Table of Contents".

As action items, we can consider:

- See how its done for Document Outline
- Model it after that
- Consider an enhancement issue to discuss expanding this to additional markup sources

ZebulanStanphill commented:
@aduth
>Did you mean the reverse of this? It is only called for dynamic blocks, and default values are not assigned for static blocks. I suspect it was likely done this way intentionally, since a static block doesn't require any further server-side processing.

Whoops, I misread the source code.

I looked a bit more into how the `level` attribute is handled on the Heading block. Apparently, whenever you select the `h2` option, the `level` attribute isn't even saved to the block comment delimiter. That definitely seems like a bug to me, because it makes it impossible to explicitly set the heading level to `2`. If the default were to ever change, it would change every Heading that is currently an `h2`.

This issue appears to not be specific to the Heading block, but rather an issue with how attribute values are handled in general when the value is the same as the default.

>Have you looked to see how the "Document Outline" considers headings? Not to say it couldn't be a good idea to explore being more open to heading tags used in non-core blocks, but it would give some idea for prior art. I'd think the two ought to be consistent in any case.

I took a look at the [Document Outline code](https://github.com/WordPress/gutenberg/blob/4857ad58c1241b3d63d21a6880c989b85746c3dc/packages/editor/src/components/document-outline/index.js). It works more or less the same as the current edit implementation of this block, filtering the list of blocks on the page for core Heading blocks. Due to the simple implementations, there doesn't appear to be any code worth extracting for reuse between the two.

In the future, however, If we go with extracting all heading elements from the markup, then I think we could probably share some of that code with the Document Outline.

>One worry I might have with detecting headings anywhere is if third-party blocks may be using heading tags in ways which don't necessarily fit in the model of a "Document Outline" or "Table of Contents".

I think that if a heading doesn't belong in a table of contents, it probably shouldn't be a heading in the first place. One exception might be using headings inside of an `article` or `aside` where the headings are not actually directly related to the current "article", but those could be filtered out by avoiding `h_` tags nested inside those elements.

On a slightly related note, I just got an idea for a future enhancement: we could add an option to the block to only show headings up to a certain level (e.g. no `h4`s through `h6`s). That would make it simple to customize how detailed you want your table of contents to be.

Anyway, I tried working around the heading level attribute issue by using `WP_Block_Type::prepare_attributes_for_render` to apply the default attribute values. However, that method has to be run using a `WP_Block_Type` instance. I tried to get one using `WP_Block_Type_Registry::get_instance()->get_registered( 'core/heading' )`, but I was getting errors because that was returning `null`; I guess that during the `render_callback`, the Heading block isn't registered or it's not possible to get a `WP_Block_Type_Registry` instance.

I thought about creating a modified version of `prepare_attributes_for_render` that used the parsed block metadata from the `block.json` file rather than a `WP_Block_Type`. However, this seemed a bit overkill when I knew that I only needed to do one thing: fill in the default level values.

So instead I went with a simple check to apply the default value if there is no value set. This uses a reference to the parsed `block.json` at the top level of the file. I could have just hardcoded the value `2` as the default, but I wanted to avoid relying on that specific value being the default.

Along the way, I discovered and fixed several bugs and learned about some really weird quirks that PHP has. I also rebased the PR and squashed a few commits.

There's still one issue left, and I'm not sure how to fix it: the anchor text can't be obtained from the Heading blocks. This is because the anchor attribute is not defined in the `block.json` metadata at all, but rather is added by a `supports` flag in `index.js`. I don't know how to get the attributes with `anchor` included, or if that's even provided in a server-side API.

aduth commented:
> If the default were to ever change, it would change every Heading that is currently an `h2`.

That's intended, and in fact largely the reason it works the way that it does, so that a change in the default value of an attribute takes effect retroactively.

> but I was getting errors because that was returning `null`

Were you only passing parsed heading blocks through to `get_registered`, or all of the results of `parse_blocks`? `parse_blocks` will include many entries with a `blockName` of `null`, specifically for content outside (between) blocks, often the whitespace of the post content.

> There's still one issue left, and I'm not sure how to fix it: the anchor text can't be obtained from the Heading blocks. This is because the anchor attribute is not defined in the `block.json` metadata at all, but rather is added by a `supports` flag in `index.js`. I don't know how to get the attributes with `anchor` included, or if that's even provided in a server-side API.

Yeah, this is something which still needs to be implemented server-side. There were some initial prototypes for this work as part of #18414.

See related:

- https://github.com/WordPress/gutenberg/pull/18414/commits/c7cded60c73e8e5c784743bc3474a0914c55928e
- https://core.trac.wordpress.org/ticket/49615

This can probably be implemented today, either by one of cherry-picking c7cded60c73e8e5c784743bc3474a0914c55928e or by proposing to patch this in core to enable default `supports`.

(cc @gziolo , since I know you've brought this up recently)

ZebulanStanphill commented:
>That's intended, and in fact largely the reason it works the way that it does, so that a change in the default value of an attribute takes effect retroactively.

But that makes no sense for the heading level option. A default value should only be applied to something that is currently `null`/`undefined` or otherwise set to some sort of "auto" value, e.g. how the text alignment options work. In the case of the heading level attribute, however, there shouldn't be an "auto" value; and it makes no sense for you to be able to explicitly set an `h3` but not an `h2`.

> Were you only passing parsed heading blocks through to `get_registered`, or all of the results of `parse_blocks`? `parse_blocks` will include many entries with a `blockName` of `null`, specifically for content outside (between) blocks, often the whitespace of the post content.

Non-Heading blocks had already been filtered out by my `block_core_table_of_contents_get_heading_blocks` function. And I don't understand what you mean by "passing [them] through to `get_registered`": that function doesn't take blocks as an argument; it takes a block type name and returns the associated `WP_Block_Type`.

>Yeah, this is something which still needs to be implemented server-side.

I guess that needs to be implemented before this block can be finished, unless we want to drop the anchor links feature for now?

aduth commented:
> > That's intended, and in fact largely the reason it works the way that it does, so that a change in the default value of an attribute takes effect retroactively.
> 
> But that makes no sense for the heading level option. A default value should only be applied to something that is currently `null`/`undefined` or otherwise set to some sort of "auto" value, e.g. how the text alignment options work. In the case of the heading level attribute, however, there shouldn't be an "auto" value; and it makes no sense for you to be able to **explicitly set an `h3` but not an `h2`**.

(Added emphasis is mine)

Your phrasing here recalled me to the issue at #7342, which essentially describes this problem. While I do think the current behavior is largely working as intended in how to treat an _implicitly undefined_ attribute value, I do sense that if a user makes an explicit choice, the serializer shouldn't go out of its way to "unset" it.

From my comment at https://github.com/WordPress/gutenberg/issues/7342#issuecomment-435371583:

>Now, if the user had taken an explicit action in choosing a value which happened to align with the block's current default at the time, I'd probably agree that this should be included in the comment demarcation, such that the choice be respected should the default be updated in the future.

I think it may be a matter of removing this code:

https://github.com/WordPress/gutenberg/blob/6c1355c0a70927a4073a8a03eb01d50f3c427f6b/packages/blocks/src/api/serializer.js#L193-L199

aduth commented:
> And I don't understand what you mean by "passing [them] through to `get_registered`": that function doesn't take blocks as an argument; it takes a block type name and returns the associated `WP_Block_Type`.

I was responding to your remark at https://github.com/WordPress/gutenberg/pull/21234#issuecomment-614047867 :

>I tried to get one using WP_Block_Type_Registry::get_instance()->get_registered( 'core/heading' ), but I was getting errors because that was returning null

My phrasing was odd, but I intended to mean the block name, where the block name of freeform content is `null`.

aduth commented:
> > Yeah, this is something which still needs to be implemented server-side.
> 
> I guess that needs to be implemented before this block can be finished, unless we want to drop the anchor links feature for now?

Yeah, it's a tough situation. For me, being able to link to the relevant heading from the table of contents is pretty critical to the block featureset, so if that depends on the availability of this, I'd consider it a blocker.

I do wonder though, is the anchor value excluded from the parsed block? I might expect it to be removed by `prepare_attributes_for_render`, but the results returned from `parse_blocks` won't have any awareness of registered blocks and their attributes schemas, so I would expect `align` to be retained until the point at which `prepare_attributes_for_render` is called.

ZebulanStanphill commented:
I decided to redo the heading extraction to use DOM manipulation and support all `h1` through `h6` elements. It now fully works in both the editor and the front-end.

There is one major remaining issue, however: saving gives an "Updating failed" notice, even though the save ia actually successful. Any ideas as to what is causing this?

ZebulanStanphill commented:
Since this is getting pretty close to ready, we should probably also start thinking about the block icon. Right now, it's just using the `list-view` dashicon, but it should probably get a unique icon that fits with the G2 style.

ZebulanStanphill commented on packages/block-library/src/table-of-contents/edit.js:
```diff
@@ -0,0 +1,89 @@
+	}
+
+	return (
+		<Block.div className={ className }>
```
So far, I've kept this as a `div` as it was in #21040. But is there any particular reason why it's a `div` in the edit implementation, or should I change it to a `nav` to match the front-end markup?

mahnunchik replied:
+1 to `nav` because it is navigation

ZebulanStanphill replied:
I have now changed it to `Block.nav`. :slightly_smiling_face:

ZebulanStanphill commented:
After some rebasing and refactoring to use the new block context system introduced in #21467, there are no longer any errors on save. I'm not exactly sure why, but it looks like everything is working now! :smiley:

mahnunchik commented:
I'm looking forward to the release of Table of Contents block 😄

SeanDS commented:
Yeah, excellent work @ZebulanStanphill! This is my #1 most wanted block in Gutenberg.

Soean commented:
Nice block, it would save us a lot of work on some sites. Here is some feedback from my first testing:

- Saving fails with 2 table of contents blocks.
- If I add the block to the Gutenberg demo post, I see php wanrings on the frontend:
   `Warning: DOMDocument::loadHTML(): Tag figure invalid in Entity, line: 8 in /var/www/src/wp-content/plugins/gutenberg/build/block-library/blocks/table-of-contents.php on line 54`
- If I add the block to the demo posts the there are noch anchor links in the table.
- Maybe we should show the placeholder component, if the post has no headings
- On some site we only want to list the h2 headlines in the table of contents block. This option would be very helpful.

ZebulanStanphill commented:
@Soean
>`Warning: DOMDocument::loadHTML(): Tag figure invalid in Entity`

Some quick research says this is caused by `loadHTML` not supporting HTML5 elements. There are 2 workarounds I've seen suggested:

- Use PHP's `strip_tags` to strip out all tags but `h1`-`h6`, `iframe`, and `template`. (The latter two have to be removed with their content, and `template` in particular has to be removed prior to parsing since it is a new element.)
- Use something like regular expressions to replace all HTML5 elements with `div`s.
- Use `libxml_use_internal_errors(true);` to prevent libxml errors from becoming visible. It's worth noting that even if a heading is nested inside an HTML5 element, the heading itself is still parsed properly and included in the table of contents.

I tried thinking of ways to use the first 2 options, but I couldn't figure out a way to use them that wasn't super complicated and fragile. I think in this case it is acceptable to simply ignore errors during the parsing, since there's always going to be a chance for invalidated markup, HTML5 support or not.

ZebulanStanphill commented:
@Soean
>Saving fails with 2 table of contents blocks.

I looked into this: it is caused by me using a function delcaration inside of another function declaration. Apparently the first declaration never goes out-of-scope. PHP's differences from JavaScript keep tripping me up! :stuck_out_tongue:

>If I add the block to the demo post the there are no anchor links in the table.

Just to be clear, the Table of Contents block does _not_ add anchor links by itself. You have to add anchors to the Heading blocks yourself. (Notably, some 3rd party Heading blocks auto-generate unique anchors; maybe we could update the core Heading block to do the same.)

I tested, and adding anchors to the headings in the demo post links them in the Table of Contents, so that appears to be working fine.

>- I cant select an empty table of contents block.
>- Maybe we should show the placeholder component, if the post has no headings

Good idea!

>On some sites we only want to list the h2 headlines in the table of contents block. This option would be very helpful.

I agree (and suggested something like this myself earlier), but I think that could be done in a follow-up PR.

ZebulanStanphill commented:
I've addressed all feedback except the show-`h2`s-only idea, which I'll leave for another PR unless others think it's a must-have. _Now_ everything should be working fine.

2 minor things left to address before this can be merged (assuming nobody finds any more bugs I missed):

- A G2-style block icon would be nice, though I'm not opposed to merging this with the current dashicon.
- Should the PHP helper function names start with an underscore? They already have PHPDoc `@access private` tags. I'm not sure what the convention is here. Also, is it okay to tag the register/render functions as private?

ZebulanStanphill commented:
Rebased to work with changes in #21868.

ZebulanStanphill commented:
I just tested this branch via gutenberg.run, and I ran into an issue with this code:

```php
mb_convert_encoding(
	'<!DOCTYPE html><html><head><title>:D</title><body>' .
		$post->post_content .
		'</body></html>',
	'HTML-ENTITIES',
	'UTF-8'
)
```
The problem is that `mb_convert_encoding` is `undefined` since it is added by a certain PHP extension not included in the gutenberg.run environment. I've replaced that code with the following, which I _think_ is equivalent. Correct me if I'm wrong.

```php
htmlspecialchars_decode(
	utf8_decode(
		htmlentities(
			'<!DOCTYPE html><html><head><title>:D</title><body>' .
				$post->post_content .
				'</body></html>',
			ENT_COMPAT,
			'UTF-8',
			false
		)
	),
	ENT_COMPAT
)
```

One question I have is whether `$post->post_content` is always encoded as UTF-8 or if it depends on the [PHP `default_charset` config option](https://www.php.net/manual/en/ini.core.php#ini.default-charset).

I've also discovered something that may be considered a blocker: the links in this block don't account for multi-page posts (posts that use the Page Break block/comment-tag). Links to headings on another page don't work. I think I can fix that.

I suppose we should also consider adding an option to only list headings for the current page.

ZebulanStanphill commented:
The block now supports paginated posts (with a few issues in the editor implementation). Links to headings on other pages will work properly. There is also a new toggle to include all headings in the post or just the ones on the current page.

The front-end (PHP) implementation should work as intended in all situations. However, it relies on multiple WordPress globals, which hopefully can be replaced by block context in the future.

The editor (JS) implementation should _almost_ always work correctly, but there are some edge cases I have been unable to account for. As far as I know, there's no way to figure out what page a block will be on in the front-end; there's no equivalent to the WordPress PHP `$page` global. Because of that, I have no good way of figuring out what page a Table of Contents block will be on, so when the "Show only headings from current page" option is enabled, it's tricky to figure out which headings should be displayed.

I'm currently calculating the page by using `getBlockIndex` and `getBlockOrder` to count how many `core/nextpage` blocks occur before the current Table of Contents block. Notably, this approach doesn't account for `<!--nextpage-->` tags in other blocks, so in some (unlikely) edge cases, the Table of Contents block will display the wrong headings in the editor.

I suspect this approach is less than optimized. Ideally, block context or some other dedicated pagination API could be used to figure out the current page. But I am currently unaware of any such methods.

If we want to get this block in sooner, I could omit the option from the initial version and just have the block always display all headings in the post. But either way, I need help to figure out how to do this better.

The wording on the new toggle is also a bit verbose, and I would appreciate any suggestions to simplify the text.

![image](https://user-images.githubusercontent.com/19592990/80720029-f27fbb80-8ac1-11ea-8629-b288e70e15f0.png)

Similarly, on the technical side, I'm not entirely sold on the attribute name I'm currently using: `showOnlyCurrentPageHeadings`.

The block also still lacks a G2 icon.

If anything in the code seems unclear, let me know and I'll add more comments.

ZebulanStanphill commented:
Fixed merge conflicts. Rebasing causes recent comments to be hidden away, so [here's a link to my previous comment](https://github.com/WordPress/gutenberg/pull/21234#issuecomment-621881956).

michelleweber commented:
I feel like you could just say "Show page headings" for the toggle label, and then "On paginated posts, only headings from the current page display."

ZebulanStanphill commented:
> I feel like you could just say "Show page headings" for the toggle label, and then "On paginated posts, only headings from the current page display."

"Show page headings" didn't feel clear enough to me. Technically, the block is always showing page headings. I did manage to come up with another toggle name that I think works better. I also updated the toggle help text and attribute name to match.

The attribute is now called `onlyIncludeCurrentPage`.

Toggle label: "Only include current page"

Help text if enabled: "Only including headings from the current page (if the post is paginated)."

Help text if disabled: "Toggle to only include headings from the current page (if the post is paginated)."

![image](https://user-images.githubusercontent.com/19592990/81139444-70f9b480-8f2b-11ea-8d39-15c3b30e4732.png)

![image](https://user-images.githubusercontent.com/19592990/81139428-60e1d500-8f2b-11ea-9c9f-cc7b2b20a54a.png)

I've also added some keywords to the block `index.js`.

Something that has just occurred to me is that the name of this block might not be the best. "Table of contents" traditionally refers to a _table_, with chapter/heading names in one column and page numbers in another. But this block doesn't use a table, nor does it list page numbers. Arguably, it's misleading to call this a "table of contents".

A lot of MediaWiki-based wikis have something similar to this block, but there it's called just "Contents":

![image](https://user-images.githubusercontent.com/19592990/81136411-21ae8680-8f21-11ea-9db8-ec0bf71508ae.png)

Additionally, in the block editor itself, we already have the Document Outline.

Perhaps this block would better be called a "Document Outline" block or a "Contents Summary" block. (I'm currently leaning towards the former.) Keep in mind that "table of contents" can still be kept as a keyword so the block shows up when you search for that term.

Side note: I don't think it's a bad idea to reuse the "document outline" term for a block. It's similar to the Block Navigation vs. the Navigation block: the term "navigation" is used by both because that's exactly what both of them are. If the Document Outline dropdown in the editor is well-named, and we're adding a block that serves a similar function, that's arguably a good reason for them to use the same terminology.

Also, since I've once again rebased the PR, [here's another link to my comment detailing the other work that I need help with on this PR](https://github.com/WordPress/gutenberg/pull/21234#issuecomment-621881956).

ZebulanStanphill commented:
Rebased to resolve merge conflicts and squashed a bunch of commits.

ZebulanStanphill commented:
Noting that this is blocked by #22874.

mahnunchik commented:
Maybe there is sense to pack it as a plugin? I really need the Table of Content block and I'd like to test it as soon as possible.

thisislawatts commented:
@ZebulanStanphill looks like the failing build here is a flake on the CI pipeline. Do you have suitable access to restart it?

Soean commented:
@thisislawatts I restarted the tests, they pass now.

thisislawatts commented:
Thanks @Soean! Now who has the capability to merge this PR? 🎉

SeanDS commented:
Wow, this is ready!? Great work @ZebulanStanphill! Back when I worked on the early version of this in #21040 I had no idea it would need so much more work from someone more experienced than me! Is there a summary anywhere of how this now works? I see a lot of new stuff.

I'm really looking forward to having this in Gutenberg. I think loads of others will appreciate it too, even if they don't know it yet! Top work.

paaljoachim commented:
Does this PR (Table of Contents block might need a shorter name) need developer documentation?

aristath commented:
Getting a fatal error on my end:
`PHP Fatal error:  Uncaught Error: Call to undefined function gutenberg_get_post_from_context() in wp-content\plugins\gutenberg\build\block-library\blocks\table-of-contents.php:361`

ZebulanStanphill commented:
@aristath Fixed (and rebased). Turns out I was still using `gutenberg_get_post_from_context()`, which got removed a while ago in favor of block context.

@paaljoachim I've tried to keep the code around this block well-commented, but if there's anything that doesn't make sense to you, please let me know.

To everyone wondering about the current status of this PR:

I've updated the PR description to better explain how the block works, as well as a description of the two major issues I see with my current implementation. Any technical feedback/reviews are very welcome, and quite possibly needed.

I should also note that there appears to be an issue with selecting the block in the editor right now. I'm pretty sure it was working before, so I assume it must be a recent regression in `master` unrelated to this PR.

aristath commented:
Testing this block it looks like after adding it I can't select it. Tried it in Chrome & FF (Ubuntu) and I can't get the block selected. So the only way to actually see the block settings, move it or delete it is to switch to the code editor, manually delete it, then go back to the visual editor, re-add it where I want it, and then it works and is selectable. It's weird and I can't exactly pinpoint where that bug is coming from, but worth mentioning in case you have any ideas :+1: 

Also I'm not sure how useful it would be to end users... If an article is long enough to require a table of contents, it's safe to assume that they'll want to make life easier for users with this block. With that in mind, maybe the list of headings should link to the actual headings? If it doesn't link to the headings somehow, then it's a simple list with no purpose other than to give an overview of the post's structure. If on the other hand it acts as a navigation, then its usefulness will increase by a lot, making navigating long posts a lot more convenient.
In the premium version of Yoast SEO we currently have a table-of-contents block that does exactly that... It goes through the article headings, if one doesn't have an `id` then one is generated, and the table of contents items link to these anchors.
I discussed this with @jdevalk and if the team is interested, we could port that functionality here.
Would we be interested in adding that functionality to the table-of-contents block?

ZebulanStanphill commented:
@aristath Headings with ids are automatically linked in the Table of Contents block. (This includes any `h1`-`h6` element with an `id` attribute... not just Heading blocks!)

Trying to auto-generate ids for headings when a Table of Contents block is inserted would be technically complicated and, in my opinion, goes against the nature of blocks being self-contained modules. However, I think it would definitely make sense for the Heading block to auto-generate its own id when inserted; but that's outside the scope of this PR.

ZebulanStanphill commented:
I made a mistake in the PR description. The specific problem with the edit implementation wasn't that I can't determine which page a heading will appear on, but rather that **I can't determine which page the Table of Contents itself will appear on**.

However, there's some good news on this front. I have revised the editor implementation so it now counts `<!--nextpage-->` tags inside Classic blocks. This means that old posts using the Classic block will work just fine with the Table of Contents block. It accomplishes this by running a string search over the entire content of every Classic block preceding the Table of Contents to find and count any `<!--nextpage-->` tags within them.

It's still not perfect, however. `<!--nextpage-->` tags in any block other than a Page Break or Classic block will still not be counted. Technically, I could fix this by searching the content of _every_ block, but I fear that would be very bad for performance, since it would mean running a regex on every block preceding the Table of Contents **every time the editor content changes**.

So what do we do? I only see two possible solutions:

1. Officially deprecate usage of `<!--nextpage-->` tags outside of a Page Break (and maybe also Classic) block.

This would make it clear that `<!--nextpage-->` tags inside something like a Paragraph block is not supported.

2. Implement a headings + document outline API that blocks can hook into, and strongly recommend that all blocks containing headings or page breaks hook into it.

Blocks could tell the editor how many page breaks and how many (and what level of) headings they contain, and this info would then be stored in a central location. The blocks could update the info they are sharing every time they are edited, or after some other trigger that they determine. For example, a Heading block could tell the editor the level and title

A block could also hook into this central data store and determine its position within it and the closest preceding heading and its level. This would allow, for example, the Table of Contents block to quickly determine which page it would be part of on the front-end. It would also allow a heading block (both the core one as well as any 3rd-party one) to determine the level of the heading immediately preceding it.

This would be far more efficient than re-scanning every block every time anything in the editor changes. It would also enable the Document Outline tool in the top bar to support 3rd party blocks (it currently does not, only including core Heading blocks). It would also allow for the heading level validator in #22650 to be completed/merged, which currently suffers from the same lack of support for 3rd party heading blocks.

The problem with the current situation is that you end up either parsing/scanning the entire content, or else you end up only supporting core blocks, with no way for 3rd party blocks to be supported.

On another note, this block currently has no icon designed for it, and I figure it should probably have one before merge, though I wouldn't be opposed to merging without one.

ZebulanStanphill commented:
I've put some more work into this PR, optimizing the code wherever I could and fixing some issues I had overlooked. I also decided to put my fears of performance issues to the test by playing around with the block to see how it performed in posts with large numbers of headings and Classic blocks. To my surprise, it actually seemed to perform just fine. So perhaps my performance concerns were overblown. However, I'd appreciate some further testing to confirm this.

I've updated the description yet again to further clarify the current state of the block. I now think the block may already be implemented well enough to merge as-is. Speaking of which, there are two remaining tasks that should be addressed either way:

- This block needs an icon designed for it.
- We should settle on whether this block should be called "Table of Contents", "Outline", or something else. It's best if we figure this out now rather than later.

paaljoachim commented:
Thanks for all your work on the Table of Contents block, Zeb!
I am thinking @jasmussen Joen will have some thoughts on an icon and name. I will also bring up this PR during the Design triage later today.

jasmussen commented:
I'm not super feeling this icon:

<img width="794" alt="Screenshot 2021-02-03 at 09 02 19" src="https://user-images.githubusercontent.com/1204802/106716288-8d339a80-65fe-11eb-93e9-e1c5365d6a37.png">

But I think it's probably good enough to try for starters:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M15.1 15.8H20v-1.5h-4.9v1.5zm-4-8.6v1.5H20V7.2h-8.9zM6 13c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 3c-.6 0-1-.4-1-1s.4-1 1-1 1 .4 1 1-.4 1-1 1zm5-3c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zM6 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z" fill="#1e1e1e"/></svg>
```

ZebulanStanphill committed:
Add unique icon.

Co-authored-by: Joen A <1204802+jasmussen@users.noreply.github.com>

ZebulanStanphill commented:
Thanks, @jasmussen! I've added the new icon. I agree that it's not perfect (the non-solid circle seems a bit weird since that's not really what lists look like, but removing it might make the result look too much like the List icon), but it'll do for now.

For those testing this PR, it's worth noting that there is currently an issue where clicking the block to select it doesn't work in some cases. It appears to be caused by a bug in `master` rather than this PR specifically, but it should probably be solved prior to merge. As a workaround while testing, you can use the List View tool to select the block.

paaljoachim commented:
This is what I see when testing with WordPress 5.7 beta 2 and Gutenberg plugin version 10.
Twenty Twenty One.
Chrome

Table of Contents is a good name for the block. The length of text in English fits in nicely.
<img width="669" alt="Screen Shot 2021-02-21 at 17 48 59" src="https://user-images.githubusercontent.com/5323259/108632104-806fcd00-746d-11eb-874d-616b6e708fd9.png">

I added a Heading h2
-Subheading h3
-Subheading h3

Added HTML anchors to them.

Backend (zoomed out)
![Screen Shot 2021-02-21 at 17 53 43](https://user-images.githubusercontent.com/5323259/108632166-c6c52c00-746d-11eb-823c-7199a9ec310f.png)

Frontend
<img width="390" alt="Screen Shot 2021-02-21 at 17 54 33" src="https://user-images.githubusercontent.com/5323259/108632188-e197a080-746d-11eb-98f0-6d7e2583e9c6.png">

The backend the h3 subheading are indented. These are not indented on the frontend.

The TOC block can be useful in multiple places such as when the time comes to recreate handbooks in Gutenberg such as https://developer.wordpress.org/block-editor/
and other areas as well.

ZebulanStanphill commented:
@paaljoachim Fixed the front-end issue. Thanks for spotting it.

mkaz commented:
The table of contents part works fine for me and looks good. I tested with various headings and levels and it worked as expected.

One issue was converting to static list wipes out all the data and it is just a generic list, not what I expected.

![convert-to-static](https://user-images.githubusercontent.com/45363/108926193-be4a2e00-75f2-11eb-8a8c-f13428e1bdb9.gif)

ZebulanStanphill committed:
Fix "Convert to static list" control.

ZebulanStanphill commented:
@mkaz Fixed. I broke it while fixing the prior issue, heh.

mkaz commented:
@ZebulanStanphill Confirmed, that fixed it.

There is some weird behavior with the editor overall selecting blocks that is making testing difficult. It is unrelated to this PR, so hard to know if you should merge before or just wait until the larger issue is resolved. I'll hold off on approving and see if it gets resolved quickly and retest again.

mkaz commented:
With https://github.com/WordPress/gutenberg/pull/29288 and https://github.com/WordPress/gutenberg/pull/29293 the selection and inserter issues seem to be cleared up, I rebased and it all tests fine now.

But in further testing, shouldn't the items in the Table of Contents being linked?

ZebulanStanphill commented:
There was an unrelated React Native test failure, so I rebased again.

@mkaz Glad to hear the weird selection issue is finally fixed! :smile:

The items in the table of contents will only be linked if the corresponding heading has an anchor id. It would be technically convoluted (and perhaps even un-block-like) for the Table of Contents block to modify other blocks on the page. For this reason, I think that the Heading block should be updated to create its own anchor id by default (rather than requiring a user to set one); but that is a task for another PR.

With that issue out of the way, there's one more thing I'd like to settle just to be completely safe:

**Should this block be called "Table of Contents" or "Document Outline"?** We already have a piece of UI in the block editor that displays a hierarchical list of headings... and it's called a "Document Outline".

![image](https://user-images.githubusercontent.com/19592990/109043085-93191a80-7696-11eb-8e9d-6067a89d3529.png)

I also wonder if "Table of Contents" might be a confusing name if people confuse it with the Table block? But on the other hand, "Table of Contents" is probably a more common term than "Document Outline".

Whatever we go with, I think we ought to use the same terminology everywhere. So if the block stays as-is, the "Document Outline" tool should be renamed. And if the tool stays as-is, the block should be renamed. What do you think?

ZebulanStanphill commented:
I just did some testing myself, and it seems the selection bug is still there, but it's a lot better now. The most reliable way I've found to reproduce it is to select the Table of Contents block, click outside the block, and then quickly click the Table of Contents block again.

Interestingly, doing the same thing on a Heading block produced a console error, though the Heading block seemed to be selected without a problem:

![image](https://user-images.githubusercontent.com/19592990/109044777-6e25a700-7698-11eb-9b54-297831837dd1.png)

Still, I think the behavior is good enough to merge now.

mkaz reviewed (approved):
Thanks for the update and explanations around links. I agree one block shouldn't alter the other and the next step to add default id for heading blocks. Is there a ticket already for this?

I prefer Table of Contents, I don't think it will be confused with Table block, and I think it is more common and what people will look for with this block. I'm indifferent on renaming the "Document Outline" part, makes sense for consistency and can also be done in another ticket.

So with all that... :ship:  :rocket:

ZebulanStanphill committed:
Tweak keywords.

ZebulanStanphill commented:
> I agree one block shouldn't alter the other and the next step to add default id for heading blocks. Is there a ticket already for this?

Yes, a pretty old one even: #6182.

I decided to make 2 last-minute changes before merge. First, I changed one of the search keywords from "outline" to "document outline" to ensure that searching that full term will bring up this block.

Second, I added @jrtashjian to the initial commit author list, since he worked on the original [Guidepost plugin](https://github.com/sortabrilliant/guidepost) that served as the basis for @ashwin-pc's PR, which in turn served as the basis for @SeanDS's PR, which served as the basis for mine.

And so, after [two and a half years of development](https://github.com/sortabrilliant/guidepost/commit/0a0f385bfee00b479dc475d09b295c9db6d55630), the Gutenberg block editor finally has a core Table of Contents block. To everyone who helped make this possible: **thank you**. :heart:

paaljoachim commented:
Thank you to everyone involved!
This is great news!

mahnunchik commented:
Two and half years of waiting! I'm looking forward to Gutenberg 10.2 🎉

jrtashjian commented:
I'm happy to see this get added into core and appreciate the kudos! Thanks all who were involved :rocket:

SeanDS commented:
Amazing work @ZebulanStanphill! I had no idea it would take so much effort to get this working, and there's no way I could have pulled it off the way you did here so I'm really happy you adopted it. Looking forward to this landing in core!

asafm7 commented:
I am using the YOAST TOC block. It is great to see it is arriving at the Core.

I encountered an issue with the YOAST block that might affect your implementation as well. I thought I might mention it:
When the _Blog Display_ is set to _Full post_ and the post is using a _More_ block - the TOC is showing on the blog/archive page but some of its links are broken, as they are relative links to HTML bookmarks that aren't actually on the page.

Example: 
https://staging.simply-hobbies.com/newsite/category/tutorials/#post-8

I guess that the best solution is: when the TOC block displays on an archive page - make its links absolute?

ZebulanStanphill commented:
@asafm7 I hadn't considered that possibility. Your guess was correct. I have created a PR to fix the issue (along with another one I discovered): #29394.

asafm7 commented:
Thanks @ZebulanStanphill ! Looking forward to the feature.

pratikkry commented:
thanks everyone

ricjcs commented:
It would be nice if it also supported numerical ordering. :)

NicoHood commented:
When will this feature hit the mainline wordpress release? Do we need to wait for a new wordpress version or can we install 10.2 earlier? The milestone is 9 days overdue, so I expect a release soon?

Soean commented:
@NicoHood 
10.2 will be released today (every 2 weeks). 
The table of contents block was removed from block registration, because there are open issues before we can release it, see #29718. So maybe in 10.3, 10.4 or later.

ZebulanStanphill commented:
I've got a refactor PR for the Table of Contents block (#29739), but I'm running into some infinite-loop issues and I need some help to fix them.

ndiego commented:
Has any further work been done on this?

paaljoachim commented:
This PR:
Autogenerate heading anchors
https://github.com/WordPress/gutenberg/pull/30825
is being worked on by @aristath Ari.

ZebulanStanphill commented:
@ndiego I was working on #29739 to overhaul the block, but I ran into some issues and haven't gotten any help to fix them yet.

phanduynam commented:
WordPress version 6.5.3 does not have a table of contents block, I recommend adding a Table of Contents block to WordPress 6.6 core to quickly create a table of contents based on all the headings on a page. From then on I won't need to install the Table of Contents Plugin anymore.
